File: ai/java/anomaly-detection/src/test/java/com/rbleggi/anomalydetection/AnomalyDetectionTest.java
package com.rbleggi.anomalydetection;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class AnomalyDetectionTest {
    private List<Transaction> historicalData;
    private Transaction normalTransaction;
    private Transaction anomalousTransaction;

    @BeforeEach
    void setUp() {
        historicalData = List.of(
            new Transaction("t1", "acc1", 150.00, "compra", 1000, "Sao Paulo"),
            new Transaction("t2", "acc1", 200.00, "compra", 2000, "Sao Paulo"),
            new Transaction("t3", "acc1", 180.00, "compra", 3000, "Curitiba"),
            new Transaction("t4", "acc1", 160.00, "compra", 4000, "Sao Paulo"),
            new Transaction("t5", "acc1", 190.00, "compra", 5000, "Belo Horizonte")
        );

        normalTransaction = new Transaction("t6", "acc1", 175.00, "compra", 6000, "Sao Paulo");
        anomalousTransaction = new Transaction("t7", "acc1", 5000.00, "compra", 7000, "Miami");
    }

    @Test
    @DisplayName("ZScoreStrategy detects anomalies based on standard deviation")
    void zScoreStrategy_detectsAnomalies() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new ZScoreStrategy(3.0));
        AnomalyResult result = system.detectAnomaly(anomalousTransaction, historicalData);

        assertTrue(result.isAnomaly());
        assertTrue(result.score() > 3.0);
        assertEquals("Z-Score", result.method());
    }

    @Test
    @DisplayName("ZScoreStrategy does not flag normal transactions")
    void zScoreStrategy_doesNotFlagNormal() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new ZScoreStrategy(3.0));
        AnomalyResult result = system.detectAnomaly(normalTransaction, historicalData);

        assertFalse(result.isAnomaly());
        assertTrue(result.score() < 3.0);
    }

    @Test
    @DisplayName("IQRStrategy detects outliers using interquartile range")
    void iqrStrategy_detectsOutliers() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new IQRStrategy(1.5));
        AnomalyResult result = system.detectAnomaly(anomalousTransaction, historicalData);

        assertTrue(result.isAnomaly());
        assertTrue(result.score() > 0);
        assertEquals("IQR", result.method());
    }

    @Test
    @DisplayName("IQRStrategy handles insufficient data")
    void iqrStrategy_handlesInsufficientData() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new IQRStrategy(1.5));
        List<Transaction> smallData = List.of(historicalData.get(0), historicalData.get(1));
        AnomalyResult result = system.detectAnomaly(normalTransaction, smallData);

        assertFalse(result.isAnomaly());
        assertTrue(result.reason().contains("insuficientes"));
    }

    @Test
    @DisplayName("MovingAverageStrategy detects deviations from recent average")
    void movingAverageStrategy_detectsDeviations() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new MovingAverageStrategy(3, 50.0));
        AnomalyResult result = system.detectAnomaly(anomalousTransaction, historicalData);

        assertTrue(result.isAnomaly());
        assertTrue(result.score() > 50.0);
        assertEquals("Moving-Average", result.method());
    }

    @Test
    @DisplayName("MovingAverageStrategy uses correct window size")
    void movingAverageStrategy_usesWindowSize() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new MovingAverageStrategy(2, 100.0));
        AnomalyResult result = system.detectAnomaly(normalTransaction, historicalData);

        assertNotNull(result);
        assertEquals("Moving-Average", result.method());
    }

    @Test
    @DisplayName("AnomalyDetectionSystem processes batch of transactions")
    void detectionSystem_processesBatch() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new ZScoreStrategy(3.0));
        List<Transaction> testSet = List.of(normalTransaction, anomalousTransaction);
        List<AnomalyResult> results = system.detectBatch(testSet, historicalData);

        assertEquals(2, results.size());
        assertFalse(results.get(0).isAnomaly());
        assertTrue(results.get(1).isAnomaly());
    }

    @Test
    @DisplayName("AnomalyDetectionSystem calculates distribution")
    void detectionSystem_calculatesDistribution() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new ZScoreStrategy(3.0));
        List<Transaction> testSet = List.of(normalTransaction, anomalousTransaction);
        Map<Boolean, Long> distribution = system.getAnomalyDistribution(testSet, historicalData);

        assertEquals(1L, distribution.get(false));
        assertEquals(1L, distribution.get(true));
    }

    @Test
    @DisplayName("AnomalyDetectionSystem filters anomalous transactions")
    void detectionSystem_filtersAnomalousTransactions() {
        AnomalyDetectionSystem system = new AnomalyDetectionSystem(new ZScoreStrategy(3.0));
        List<Transaction> testSet = List.of(normalTransaction, anomalousTransaction);
        List<Transaction> anomalies = system.getAnomalousTransactions(testSet, historicalData);

        assertEquals(1, anomalies.size());
        assertEquals(anomalousTransaction.id(), anomalies.get(0).id());
    }
}

File: kotlin/anomaly-detection/src/test/kotlin/com/rbleggi/anomalydetection/AnomalyDetectionTest.kt
package com.rbleggi.anomalydetection

import kotlin.test.*

class AnomalyDetectionTest {

    private val normalHistory = listOf(
        Transaction("h1", 250.0, "Account", 1000),
        Transaction("h2", 300.0, "Account", 2000),
        Transaction("h3", 280.0, "Account", 3000),
        Transaction("h4", 310.0, "Account", 4000),
        Transaction("h5", 290.0, "Account", 5000)
    )

    private val normalTransaction = Transaction("t1", 295.0, "Account", 6000)
    private val anomalyTransaction = Transaction("t2", 5000.0, "Account", 7000)

    @Test
    fun `ZScoreStrategy detects normal transaction`() {
        val strategy = ZScoreStrategy(threshold = 3.0)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertFalse(result.isAnomaly)
        assertEquals("Z-Score", result.method)
    }

    @Test
    fun `ZScoreStrategy detects anomaly`() {
        val strategy = ZScoreStrategy(threshold = 2.0)
        val result = strategy.detect(anomalyTransaction, normalHistory)

        assertTrue(result.isAnomaly)
        assertTrue(result.score > 2.0)
    }

    @Test
    fun `ZScoreStrategy respects threshold`() {
        val lowThreshold = ZScoreStrategy(threshold = 1.0)
        val highThreshold = ZScoreStrategy(threshold = 5.0)

        val resultLow = lowThreshold.detect(normalTransaction, normalHistory)
        val resultHigh = highThreshold.detect(anomalyTransaction, normalHistory)

        assertTrue(resultLow.score < 1.0 || resultLow.isAnomaly)
        assertTrue(resultHigh.score > 0.0)
    }

    @Test
    fun `IQRStrategy detects normal transaction`() {
        val strategy = IQRStrategy(multiplier = 1.5)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertFalse(result.isAnomaly)
        assertEquals("IQR", result.method)
    }

    @Test
    fun `IQRStrategy detects anomaly`() {
        val strategy = IQRStrategy(multiplier = 1.5)
        val result = strategy.detect(anomalyTransaction, normalHistory)

        assertTrue(result.isAnomaly)
        assertTrue(result.score > 0.0)
    }

    @Test
    fun `IQRStrategy calculates quartiles correctly`() {
        val strategy = IQRStrategy(multiplier = 1.5)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertTrue(result.details.contains("IQR"))
        assertTrue(result.details.contains("limites"))
    }

    @Test
    fun `MovingAverageStrategy detects normal transaction`() {
        val strategy = MovingAverageStrategy(windowSize = 5, threshold = 2.0)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertFalse(result.isAnomaly)
        assertEquals("Moving Average", result.method)
    }

    @Test
    fun `MovingAverageStrategy detects anomaly`() {
        val strategy = MovingAverageStrategy(windowSize = 5, threshold = 1.0)
        val result = strategy.detect(anomalyTransaction, normalHistory)

        assertTrue(result.isAnomaly)
        assertTrue(result.score > 1.0)
    }

    @Test
    fun `MovingAverageStrategy respects window size`() {
        val smallWindow = MovingAverageStrategy(windowSize = 3, threshold = 2.0)
        val largeWindow = MovingAverageStrategy(windowSize = 5, threshold = 2.0)

        val resultSmall = smallWindow.detect(normalTransaction, normalHistory)
        val resultLarge = largeWindow.detect(normalTransaction, normalHistory)

        assertNotNull(resultSmall.details)
        assertNotNull(resultLarge.details)
    }

    @Test
    fun `AnomalyDetector detects single transaction`() {
        val detector = AnomalyDetector(ZScoreStrategy())
        val result = detector.detect(normalTransaction, normalHistory)

        assertFalse(result.isAnomaly)
    }

    @Test
    fun `AnomalyDetector detects batch of transactions`() {
        val detector = AnomalyDetector(ZScoreStrategy())
        val transactions = listOf(normalTransaction, anomalyTransaction)
        val results = detector.detectBatch(transactions, normalHistory)

        assertEquals(2, results.size)
    }

    @Test
    fun `AnomalyDetector counts anomalies`() {
        val detector = AnomalyDetector(ZScoreStrategy(threshold = 2.0))
        val transactions = listOf(normalTransaction, anomalyTransaction)
        val count = detector.anomalyCount(transactions, normalHistory)

        assertTrue(count >= 0 && count <= 2)
    }

    @Test
    fun `AnomalyDetector calculates false positive rate`() {
        val detector = AnomalyDetector(ZScoreStrategy(threshold = 1.0))
        val transactions = listOf(normalTransaction, anomalyTransaction)
        val knownAnomalies = setOf("t2")

        val fpr = detector.falsePositiveRate(transactions, normalHistory, knownAnomalies)

        assertTrue(fpr >= 0.0 && fpr <= 1.0)
    }

    @Test
    fun `Transaction data class stores all fields`() {
        val transaction = Transaction("id", 100.0, "account", 1000L)

        assertEquals("id", transaction.id)
        assertEquals(100.0, transaction.amount)
        assertEquals("account", transaction.account)
        assertEquals(1000L, transaction.timestamp)
    }

    @Test
    fun `AnomalyResult contains all fields`() {
        val result = AnomalyResult(true, 3.5, "Z-Score", "details")

        assertTrue(result.isAnomaly)
        assertEquals(3.5, result.score)
        assertEquals("Z-Score", result.method)
        assertEquals("details", result.details)
    }

    @Test
    fun `ZScoreStrategy provides details`() {
        val strategy = ZScoreStrategy(threshold = 3.0)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertTrue(result.details.contains("Z-Score"))
        assertTrue(result.details.contains("threshold"))
    }

    @Test
    fun `IQRStrategy provides details`() {
        val strategy = IQRStrategy(multiplier = 1.5)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertTrue(result.details.contains("IQR"))
    }

    @Test
    fun `MovingAverageStrategy provides details`() {
        val strategy = MovingAverageStrategy(windowSize = 5, threshold = 2.0)
        val result = strategy.detect(normalTransaction, normalHistory)

        assertTrue(result.details.contains("Desvio"))
        assertTrue(result.details.contains("media movel"))
    }
}
